### PR TITLE
Prohibit bad --enable-linux-netfilter combinations

### PR DESCRIPTION
**Pull request by rousskov**
Since 2009 commit 51f4d36b, Squid declared that "all transparent build
options are independent, and may be used in any combination". That
declaration was accurate from a "successful build" point of view, but
Ip::Intercept::LookupNat() (and its precursors) started mishandling at
least two combinations of options as detailed below.

LookupNat() tries up to four lookup methods (until one succeeds):

1. NetfilterInterception(): --enable-linux-netfilter; SO_ORIGINAL_DST
2. IpfwInterception(): --enable-ipfw-transparent; getsockname()
3. PfInterception(): --enable-pf-transparent; getsockname() or /dev/pf
4. IpfInterception(newConn): --enable-ipf-transparent; ioctl(SIOCGNATL)

The first method -- NetfilterInterception() -- fails to look up the true
destination address of an intercepted connection when, for example, the
client goes away just before the lookup. In those (relatively common in
busy environments) cases, the intended destination address cannot be
obtained via getsockname(), but LookupNat() proceeds calling other
methods, including the two methods that may rely on getsockname().

Methods 2 and 3 may rely on a previous getsockname() call to provide
true destination address, but getsockname() answers are not compatible
with what NetfilterInterception() must provide -- the destination
address returned by getsockname() is Squid's own http(s)_port address.
When Squid reaches problematic code now encapsulated in a dedicated
UseInterceptionAddressesLookedUpEarlier() function, Squid may end up
connecting to its own http(s)_port! Such connections may cause
forwarding loops and other problems. In SslBump deployments, these loops
form _before_ Forwarded-For protection can detect and break them.

These problems can be prevented if an admin does not enable incompatible
combinations of interception lookup methods. The relevant code is
correctly documented as "Trust the user configured properly", but that
statement essentially invalidates our "may be used in any combination"
design assertion and leads to runtime failures when user configured
improperly. Those runtime failures are difficult to triage because they
lack signs pointing to a build misconfiguration.

This change bans incompatible NetfilterInterception()+getsockname()
combinations at compile time: Squid ./configured with
--enable-linux-netfilter cannot use --enable-ipfw-transparent or
(--enable-pf-transparent --without-nat-devpf).

TODO: Ban incompatible combinations at ./configure time as well!

We have considered and rejected an alternative solution where all
./configure option combinations are still allowed, but LookupNat()
returns immediately on NetfilterInterception()/SO_ORIGINAL_DST failures.
That solution is inferior to build-time bans because an admin may think
that their Squid uses other configured lookup method(s) if/as needed,
but Squid would never reach them in --enable-linux-netfilter builds.

The only viable alternative is to specify lookup methods in squid.conf,
similar to the existing "tproxy" vs. "intercept" http(s)_port modes. In
that case, squid.conf will be validated for incompatible method
combinations (if combinations are supported at all), and LookupNat()
will only use configured method(s).

### CODE REVIEW

File: src/ip/Intercept.cc [hunk 1]
```diff
@@ -182,18 +182,32 @@ bool
...
 #endif
 }
 
+/// Assume that getsockname() has been called already and provided the necessary
+/// TCP packet details. There is no way to identify whether they came from NAT.
+/// Trust the user configured properly.
+bool
+Ip::Intercept::UseInterceptionAddressesLookedUpEarlier(const char * const caller, const Comm::ConnectionPointer &newConn)
+{
+    // paranoid: ./configure should prohibit these combinations
```
Comment (rousskov) [inline src/ip/Intercept.cc:198]: This is a _draft_ PR, in part, because this ./configure-time prohibition has not been implemented yet.

**Review by rousskov — COMMENTED** [review]
This _draft_ PR was posted to ask the following question: Should we finish this implementation (by adjusting configure.ac accordingly) OR switch to the "viable alternative" outlined at the end of the current PR description: Adjust http(s)_port options to specify which intended destination lookup method(s) Squid should use for a specific intercepting port?

Needless to say, if you find some fundamental problems with this draft PR logic or implementation, please detail them as well!

**Comment by yadij** [comment]
IMO the `./configure` prohibition will not work due to cross-build needs.

FTR, The original long-term intention was to migrate back to `transparent` (meaning TRPOXY equivalent) and get rid of the NAT support. However PF and IPFW changes have since made that problematic as you found.

IMO, we should now put the NAT logic into a loadable module. Alternatively add more "mode" flags named matching the firewall configuration; eg "tproxy" for Netfilter,  "divert-natd" for IPFW, `rdr-to` for PF, etc

**Review by yadij — APPROVED** [review]

**Comment by yadij** [comment]
FTR, I already cleared this for merge as-is.   The issues were/are with the alternative future ideas proposed.

**Comment by yadij** [comment]
`
../../../../src/ip/Intercept.cc:203:19: error: static assertion failed
  203 |     static_assert(!"--enable-linux-netfilter is incompatible with --enable-ipfw-transparent");
`

You will need to remove that option from `test-suite/buildtests/layer-02-maximus.opts`

**Comment by rousskov** [comment]
> Alex: Q1: Do you want this "minimal" PR merged (after we fix tests, of course) before we change squid.conf aspects?

> Amos: I already cleared this for merge as-is.

Thank you; I am interpreting that clearance as a "yes" answer to Q1.

@eduard-bagdasaryan, please fix the tests (Amos' [hint](https://github.com/squid-cache/squid/pull/1893#issuecomment-2382504302) may be the best starting point).

@yadij, if you can answer [Q2](https://github.com/squid-cache/squid/pull/1893#issuecomment-2340800801), please do so. That answer will not affect this "minimal" PR, but it will help shape followups (if any).

**Pull request closed without merging**